Tuning of individual kernel threads

adriaan42 opened the pull request:
In combination with #596 and #580, this PR implements the third feature needed to dynamically tune all relevant aspects of a realtime application using a dedicated HW device (typically a NIC).

Two things might need some discussion:

* In my implementation I kept the basic idea that one instace can cover a number of different groups of threads. That makes it easy to migrate from the current `scheduler` plugin, but means we still need `_has_dynamic_options`, which is marked as a hack in plugins/base.py. The alternative would be to have one plugin instance per "group", which would make the profiles much longer.
  ```ini
  group.ktimers=0:f:2:*:^\[ktimers
  ```
  would become something like
  ```ini
  [kthread_ktimers]
  type=kthread
  regex=^ktimers
  policy=fifo
  sched_prio=2
  affinity=*
  ```
* I copied the approach of using perf to monitor for creation of new threads. That means that when running both the `scheduler` plugin and the `kthread` plugin, we'd have two threads doing the same thing. For my applications that's not a problem, because I no longer use the `scheduler` plugin at all:
  * `scheduler` handles three things: IRQ affinities, kernel threads, and userland threads
  * For IRQ affinities I can use the `irq` plugin
  * For kernel threads I can use `kthread`
  * For userland threads I use systemd and cgroupv2, and I don't want TuneD to touch them

yarda commented on tuned/plugins/plugin_kthread.py:
```diff
@@ -0,0 +1,561 @@
+		self._kthread_scan(initial=True)
+
+	@classmethod
+	def _get_config_options(self):
```
Please fix. The base plugin should be also fixed, but this should be done in new PR.

adriaan42 replied:
Done... one of these days I need to learn how to run those checks locally before I push... :)

yarda commented on tuned/plugins/plugin_kthread.py:
```diff
@@ -0,0 +1,561 @@
+	# command definitions: entry point for device tuning
+	#
+	@command_custom("_instance", per_device=True)
+	def _instance_kthread(self, start, value, device, verify, ignore_missing):
```
Please fix.

yarda commented:
> but means we still `need _has_dynamic_options`, which is marked as a hack in plugins/base.py. 

It's OK for me, in long-term it's a candidate for rewrite/refactor, but there are other plugins using it as well. We will probably keep the idea and if we change the implementation, this could be then updated in all affected plugins the same way.

yarda commented:
Regarding the cgroups, there is support for cgroups v1 in the scheduler plugin and we would also like to add support for the v2 for completeness. It could be useful for somebody.

It's OK if you are not using some plugin. We even wanted to add global configuration option allowing selective disablement of specific plugins in the stock profiles.

adriaan42 commented on tuned/plugins/plugin_kthread.py:
```diff
@@ -0,0 +1,561 @@
+	"""
+	def __init__(self, monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables):
+		super(KthreadPlugin, self).__init__(monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables)
+		self._has_dynamic_options = True
```
This pattern is used in many plugins, and I don't see an easy way around it without larger refactorings.

adriaan42 commented:
Update:

* rebased on latest `main`
* added check of perf events, according to finding by @zacikpa in #691

@yarda did you already have a chance to review this PR?

adriaan42 commented:
> Wouldn't be better to use one group per instance? I.e. the same affinity/sched_opts setting for individual instance, e.g.:

I thought about this when I wrote the plugin, and didn't like it because

* I like the "default" of having only one instance per plugin in the profile, unless doing some specific optimizations. That instance has the name of the plugin, so `type=` is not needed. In the current profiles only few very specific cases need multiple instances (e.g. ThunderX).
* The approach with multiple instances would make the common case (what's currently used in the `realtime` profiles) much more complex to express.

I find the current format in the `scheduler` plugin quite clear, except for minor points I've changed:

* unify policy and priority into one option (`f50` instead of `f:50`) because setting those separately makes no sense to me
* remove the braces `[]` from thread names (the `scheduler` plugin adds them to identify kernel threads, but here we have only kernel threads, so no need for this)

> ```
> [ksoftirqd]
> type=kthread
> devices_udev_regex=^\[ksoftirqd
> setting=0:f:2:*
> ```

Also re-using the existing `devices_udev_regex` is not very clean, as we're not dealing with udev Devices here. So the setting we're using in the profiles would be "wrong", and internally we'd need to create `pyudev.Device` objects just to use the existing mechanism.

yarda commented:
Here in one instance there can be multiple "devices" with multiple different tuning. I don't like that it breaks the used logic concept that in each instance there are "devices" with the same tuning. I am afraid that it could lead to multiple problems later (e.g. when moving "devices" between instances through the API).

But maybe it's just my personal preference, @zacikpa, @jmencak what's your opinion?

zacikpa commented:
I'm personally fine with the implementation as it is now (since it's inspired by the `groups.*` implementation in the `scheduler` plugin), but I would not refer to the kthreads as devices anywhere in the code/comments. In other words, the new plugin would not support "devices" in the TuneD sense, similarly to the scheduler plugin.

yarda commented:
No device plugin, like e.g. the sysfs plugin, this may work.

adriaan42 commented:
> No device plugin, like e.g. the sysfs plugin, this may work.

One of the main points of the new plugin is to allow dynamic changes of the tuning (through the `instance_[create|destroy]` dbus calls). That only works on device-plugins (based on `hotplug.Plugin`).

zacikpa commented:
Huh, taking back what I said [here](https://github.com/redhat-performance/tuned/pull/596#pullrequestreview-1938347971), I say we should adjust `instance_create` and `instance_destroy` to work with any plugin, not just child classes of `hotplug.Plugin`.

IIUC, you can't and don't plan to use `instance_acquire_devices` with this plugin, am I right? (Because it does not support what TuneD calls "devices").

adriaan42 commented:
> > Huh, taking back what I said [here](https://github.com/redhat-performance/tuned/pull/596#pullrequestreview-1938347971), I say we should adjust `instance_create` and `instance_destroy` to work with any plugin, not just child classes of `hotplug.Plugin`.
> 
> The change itself is probably simple, but I'm not sure what would happen for any of the existing non-hotplug Plugins if multiple instances were created. I expect those instances would just interfere with each other and break things.
> 
> > IIUC, you can't and don't plan to use `instance_acquire_devices` with this plugin, am I right? (Because it does not support what TuneD calls "devices").
> 
> `instance_acquire_devices` could be used, but it does not make much sense in this case. (One could force the transfer of a "device" to an instance that does not match it, but then no tuning would be applied).
> 
> Treating kthreads as "devices" lets me reuse a lot of nice infrastructure. But it would of course be possible to do this all within the plugin.

@zacikpa I just pushed a draft that bases the plugin on `base.Plugin` instead of `hotplug.Plugin`. The changes in `controller.py` probably need some more work.

yarda commented:
@adriaan42 could you please squash the commits? Multiple commits each introducing separate logical change are OK, but multiple commits on the same file (e.g. `plugin_kthread.py`) reverting/changing things are making it a bit confusing.

Regarding the `controller.py` changes, I am afraid this could break something - this needs thorough testing at least. I am not brave enough to push it into the release now.

@zacikpa your opinion?

yarda commented:
Or maybe split this PR to two PRs. The commit adding switch for disabling kthreads can be merged immediately and removing it from this PR could make rebasing easier.

adriaan42 commented:
> @adriaan42 could you please squash the commits? Multiple commits each introducing separate logical change are OK, but multiple commits on the same file (e.g. `plugin_kthread.py`) reverting/changing things are making it a bit confusing.

I pushed the changes as separate commit as part of the discussion, to show required changes and maybe get some feedback (confirmation that this might be the way to go before I spend effort to finalize the imlementation). Will post an update soon.

> Regarding the `controller.py` changes, I am afraid this could break something - this needs thorough testing at least. I am not brave enough to push it into the release now.

I agree. I'm happy to look into better solutions and possible side-effects, if we agree that "allow dynamic instances of non-device plugins" is reasonable and has a chance to be integrated. Again, trying to avoid effort for something that might never get merged...

> Or maybe split this PR to two PRs. The commit adding switch for disabling kthreads can be merged immediately and removing it from this PR could make rebasing easier.

Sure! -> #734

adriaan42 commented:
Update:

* rebase following the latest release
* implemented seamless transition of kthreads between plugin instances (similar to #750, but here the plugin needs to handle everything as part of "static" tuning, because we're no longer a "device" plugin)

adriaan42 commented on tuned/plugins/plugin_kthread.py:
```diff
@@ -0,0 +1,736 @@
+	def _instance_release_kthreads(self, instance):
+		"""release all kthreads from an instance"""
+		with self._lock:
+			free_kthreads = instance._kthreads_assigned | instance._kthreads_processed
+			# first the ones now claimed by other instances
+			for other_instance in self._instances.values():
+				if other_instance == instance:
+					continue
+				transfer_kthreads = self._get_matching_kthreads(other_instance, free_kthreads)
+				for pid in list(transfer_kthreads):
+					self._instance_transfer_kthread(instance, other_instance, pid)
+					free_kthreads.remove(pid)
```
`self._instances` is already a `collections.OrderedDict` sorted by `instance.priority` (ascending, i.e., highest-priority-first since lower number = higher priority). This is maintained on every insertion in base.py:

```python
self._instances = collections.OrderedDict(sorted(self._instances.items(), key=lambda x: x[1].priority))
```

The loop in `_instance_release_kthreads` that iterates `self._instances.values()` already processes instances in priority order (highest priority first). No change needed.

adriaan42 commented on tuned/plugins/plugin_kthread.py:
```diff
@@ -0,0 +1,764 @@
+
+		global have_perf
+		if have_perf:
+			try:
+				self._perf_setup()
+			except Exception as e:
+				try:
+					self._perf_shutdown()
+				except Exception:
+					pass
+				have_perf = False
```
The `except: pass` around `_perf_shutdown()` in the `__init__` error path (lines 145-147) is a **best-effort cleanup** after a failed initialization. This is the standard pattern for graceful degradation:

1. `_perf_setup()` fails → partial resources may exist
2. `_perf_shutdown()` attempts to close any partially-opened perf file descriptors (note it already guards with `getattr(self, "_evlist", None)`)
3. `have_perf = False` is set → plugin operates in degraded mode (tunes existing kthreads but can't monitor for new ones)

Re-raising a shutdown exception here would **crash the entire plugin** instead of allowing it to degrade gracefully. The claim that this "prevents `cleanup()` from retrying" is incorrect — `cleanup()` checks `if have_perf:` and should only call `_perf_shutdown()` when perf was *successfully* initialized. If setup failed, there's nothing meaningful for `cleanup()` to retry.

adriaan42 committed:
feat: add plugin_kthread

Signed-off-by: Adriaan Schmidt <adriaan.schmidt@siemens.com>

adriaan42 commented:
@coderabbitai resume

adriaan42 committed:
fix: address review finding
